### PR TITLE
Feat #208 Cluster configuration

### PR DESCRIPTION
**Pull request by thomas-tomlinson**
Addresses the concerns in issues #208.

### CODE REVIEW

**Comment by alaa-bish** [comment]
we have tried to create some update tests but seems the code is not working properly , 
update scenarios :
update redundancy_factor > the request is passed with no error but the value keep same it's not updated
update supported_information_verbosity > the request is passed with no error but the value keep same it's not updated
there is no way to update cluster name 
update timezone sometimes working and sometimes not wotking

File: .gitignore [hunk 1]
```diff
@@ -1 +1,3 @@
 nutanix-ncp-*
+tests/
```
Comment (bhati-pradeep) [inline .gitignore:2]: Can you revert this changes. Tests contributions are very often by others.

File: plugins/module_utils/prism/clusters.py [hunk 2]
```diff
@@ -5,13 +5,104 @@
...
+            "supported_information_verbosity": self._build_spec_supported_information_verbosity,
+            "redundancy_factor": self._build_spec_redundancy_factor,
+            "network": self._build_spec_network,
+        }
 
+    def _get_default_spec(self):
+        return deepcopy(
+            {
+                "api_version": "3.1.0",
+                "metadata": {"kind": "cluster", "spec_version": 2},
```
Comment (bhati-pradeep) [inline plugins/module_utils/prism/clusters.py:27]: Spec versions is not required, we generally use metadata coming from get API when there is only update functionality

File: plugins/module_utils/prism/clusters.py [hunk 3]
```diff
@@ -5,13 +5,104 @@
...
+                    "resources": {
+                        "config": {
+                            "authorized_public_key_list": []
+                        }
+                    }
+                }
+            }
+        )
+
+    def _build_spec_redundancy_factor(self, payload, redundancy_factor):
```
Comment (bhati-pradeep) [inline plugins/module_utils/prism/clusters.py:39]: if redundancy_factor is none from user then it would never reach here. As None values are omitted in Entity.get_spec() and even from ansible module.params. This applies for all fields from input so we can change accordingly in all build spec methods

Comment (bhati-pradeep) [reply plugins/module_utils/prism/clusters.py:39]: Although we can restrict it from passing 0

File: plugins/module_utils/prism/clusters.py [hunk 4]
```diff
@@ -5,13 +5,104 @@
...
+                payload["spec"]["resources"]["network"][single_entry] = network[single_entry]
+
+        # special snowflake with keyword option collision.  smtp server has an argument of "type"
+        # which is a keyword in the ansible argument spec.  this will remap the value back to type in the
+        # spec file is specified.  this will change server_type back to type.  the PLAIN default is
+        # used here if nothing is found in the config or existing spec file
+        if "smtp_server" in network:
+            if "server_type" in network["smtp_server"]:
+                network["smtp_server"]["type"] = network["smtp_server"]["server_type"]
+                del network["server_type"]
```
Comment (bhati-pradeep) [inline plugins/module_utils/prism/clusters.py:76]: Why would network have entry with key "server_type" ?

File: plugins/module_utils/prism/clusters.py [hunk 5]
```diff
@@ -5,13 +5,104 @@
...
+            elif "type" in payload["spec"]["resources"]["network"]["smtp_server"]:
```
Comment (bhati-pradeep) [inline plugins/module_utils/prism/clusters.py:77]:
```suggestion
            elif "type" in (payload["spec"]["resources"]["network"]).get("smtp_server"):
```
Get API response might not have stmp server entry. Failed for me in local in this case.

File: plugins/module_utils/prism/clusters.py [hunk 6]
```diff
@@ -5,13 +5,104 @@
...
+                network["smtp_server"]["type"] = payload["spec"]["resources"]["network"]["smtp_server"]["type"]
+            else:
+                network["smtp_server"]["type"] = "PLAIN"
+
+            payload["spec"]["resources"]["network"]["smtp_server"] = network["smtp_server"]
+
+        return payload, None
+
+    def get_current_spec(self):
```
Comment (bhati-pradeep) [inline plugins/module_utils/prism/clusters.py:86]: getting cluster uuid can be done by existing get_cluster_uuid() helper and the logic to create old spec from cluster info can be shifted to ntnx_clusters.py which will call get_cluster_uuid()

File: plugins/modules/ntnx_clusters.py [hunk 7]
```diff
@@ -0,0 +1,522 @@
...
+        nameserver=dict(type="str"),
+        domain_credentials=dict(
+            type="dict", options=dict(
+                username=dict(type="str"),
+                password=dict(type="str", no_log=True),
+            )
+        )
+    )
+
+    http_proxy_list_spec = dict(
```
Comment (bhati-pradeep) [inline plugins/modules/ntnx_clusters.py:370]: Incase some field is required in subspec of http_proxy_list_spec, domain_server_spec, etc. mark it as required=True

File: plugins/modules/ntnx_clusters.py [hunk 8]
```diff
@@ -0,0 +1,522 @@
...
+                credentials=dict(
+                    type="dict", options=dict(
+                        username=dict(type="str"),
+                        password=dict(type="str", no_log=True),
+                    ),
+                ),
+                proxy_type_list=dict(type="list", elements="str"),
+            )
+        ),
+        server_type=dict(type="str"),
```
Comment (bhati-pradeep) [inline plugins/modules/ntnx_clusters.py:415]: If this field only allows few values then we can mention choices, for ex. choices=["PLAIN", "SMTP"]

File: plugins/modules/ntnx_clusters.py [hunk 9]
```diff
@@ -0,0 +1,522 @@
...
+    resp = cluster.update(spec, uuid)
+    result["task_uuid"] = resp["status"]["execution_context"]["task_uuid"]
+
+    if module.params.get("wait"):
+        wait_for_task_completion(module, result)
+        resp = cluster.read(uuid)
+        result["response"] = resp
+
+    # compare the specs before and after.  currently only way to determine if something changed
+    if old_spec["spec"] != result["response"]["spec"]:
```
Comment (bhati-pradeep) [inline plugins/modules/ntnx_clusters.py:485]: For idempotency checks, we compare old spec and update spec before making api call to update. If there is no difference then we skip the update call. This is how we make our modules idempotent. Please refer other modules for same. You can compare the specs directly if they dont required indepth checks. It should not make update call if old spec and update spec are equivalent. Ex: https://github.com/nutanix/nutanix.ansible/blob/b82b58d0a767e821f64107c2f363cebfaa892822/plugins/modules/ntnx_protection_rules.py#L700

Comment (bhati-pradeep) [reply plugins/modules/ntnx_clusters.py:485]: Post update call changed will be always true incase idempotency checks are placed before calls

File: plugins/modules/ntnx_clusters.py [hunk 10]
```diff
@@ -0,0 +1,522 @@
...
+    result["response"] = resp
+
+
+def run_module():
+    module = BaseModule(
+        argument_spec=get_module_spec(),
+        supports_check_mode=True,
+    )
+    remove_param_with_none_value(module.params)
+    result = {
```
Comment (bhati-pradeep) [inline plugins/modules/ntnx_clusters.py:504]: Add cluster_uuid as well. Its usefull when results of this module's task are used in other tasks in playbook

File: plugins/modules/ntnx_clusters.py [hunk 11]
```diff
@@ -0,0 +1,522 @@
...
+        key=dict(type="str", no_log=False),
+    )
+
+    module_args = dict(
+        cluster=dict(
+            type="dict", options=entity_by_spec, mutually_exclusive=mutually_exclusive, required=True,
+        ),
+        authorized_public_key_list=dict(type="list", elements="dict", options=authorized_public_key_list_spec),
+        timezone=dict(type="str"),
+        supported_information_verbosity=dict(type="str"),
```
Comment (bhati-pradeep) [inline plugins/modules/ntnx_clusters.py:456]: I beleive this only accepts 4 values. We can put them in choices = []

File: plugins/module_utils/prism/clusters.py [hunk 12]
```diff
@@ -5,13 +5,104 @@
...
+            if not uuid:
+                error = "Cluster {0} not found.".format(name)
+                return None, error
+        elif "uuid" in config:
+            uuid = config["uuid"]
+        else:
+            error = "Config {0} doesn't have name or uuid key".format(config)
+            return None, error
+        cluster_info = self.read(uuid)
+        old_spec = {}
```
Comment (bhati-pradeep) [inline plugins/module_utils/prism/clusters.py:100]: In other way we can remove status from cluster_info which will eventually be old_spec ?

**Review by bhati-pradeep — CHANGES_REQUESTED** [review]

File: plugins/modules/ntnx_clusters.py (hunk 11 above)
Comment (bhati-pradeep) [reply plugins/modules/ntnx_clusters.py:456]: BASIC_PLUS_CORE_DUMP, ALL, BASIC, NOTHING

**Review by bhati-pradeep — COMMENTED** [review]
Please add field to update cluster name as well, cluster API supports as per documentation : https://www.nutanix.dev/api_references/prism-central-v3/#/90cf8c01b8623-update-a-existing-cluster

File: plugins/modules/ntnx_clusters.py (hunk 11 above)
Comment (bhati-pradeep) [reply plugins/modules/ntnx_clusters.py:456]: Refer https://www.nutanix.dev/api_references/prism-central-v3/#/90cf8c01b8623-update-a-existing-cluster

**Comment by thomas-tomlinson** [comment]
it looks like this work has been completed and merged into the main branch.  as i no longer work for a company using this combination of tech, i'm going to close this PR.

**Comment by bhati-pradeep** [comment]
Thanks for informing, this work is not closed yet. We will be targetting Clusters configuration with v4 API integration releases.
